elastic/kibana
# [Security solution] Bedrock token tracking - dashboard link added to connector UI

## Summary

I ended up adding token tracking for the bedrock `invokeAI` method since [LangChain streaming](https://github.com/elastic/kibana/pull/171455) is at risk. I will remove it if I am able.

The majority of this PR is making adjustments to the GenAI dashboard code to be able to be used for both types of GenAI connectors.

### To test

1. Create Bedrock and OpenAI connectors both in the UI and in `kibana.dev.yml` with preconfigured actions. [Credentials are here](https://p.elstc.co/paste/RnjcCG2T#zs2MS26xaBSQxcjF65B9e+YXmI0LZuOvehv6LdrdYse) (I currently only have Azure OpenAI and Bedrock credentials to test with).
2. With each connector:
  **a.** send a test message through the Test connector view (tests token tracking on  `run`/`test` subactions). 
  **b.** Send a security AI assistant message with knowledge base turned off (tests token tracking on `invokeStream` subaction). 
  **c.** Send a security AI assistant message with knowledge base turned off (tests token tracking on  `invokeAI` subaction).
4. From the Edit connector view of each connector, click through to the dashboard. For example, bedrock:
<img width="1440" alt="Screenshot 2023-11-28 at 5 07 11 PM" src="https://github.com/elastic/kibana/assets/6935300/0aa8c14d-582e-436d-8c20-5fc2e06acede">

5. You should be taken to a dashboard with an id of either `generative-ai-token-usage-bedrock-default` or `generative-ai-token-usage-openai-default`. The dashboard should be named appropriately ("Amazon Bedrock Token Usage" or "OpenAI Token Usage"). You should see data in the 2 Lens visualizations.

6. Use the "Explore data in discover" button from the "Prompt + Completion Tokens per User" visualization to open Discover
<img width="1406" alt="Screenshot 2023-11-28 at 5 23 13 PM" src="https://github.com/elastic/kibana/assets/6935300/b9c757b8-3c4b-4d2d-af1e-98b264139055">

7. You should see token tracking for 3 events (a, b, and c of step 2)

<img width="1432" alt="Screenshot 2023-11-28 at 5 25 22 PM" src="https://github.com/elastic/kibana/assets/6935300/1a83d48f-4b38-4170-8314-30d7f6fe2df4">
